File: bigdoors_bp/scripts/domain/MaterialRegistry.js
import {
  MATERIAL_INDEX,
  MATERIAL_GEOMETRY_CLASS,
  GEOMETRY_CLASS_FENCE,
  GEOMETRY_CLASS_BARS,
  GEOMETRY_CLASS_PANE,
  DIR_OFFSETS,
  panelBlockIdForGeoClass,
} from "../util/Constants.js";

/** @type {Map<string, number>} typeId → index */
const _typeIdToIndex = new Map();
MATERIAL_INDEX.forEach((typeId, index) => {
  _typeIdToIndex.set(typeId, index);
});

export function indexForTypeId(typeId) {
  const idx = _typeIdToIndex.get(typeId);
  return idx !== undefined ? idx : -1;
}

export function typeIdForIndex(index) {
  return MATERIAL_INDEX[index];
}

export function geometryClassForMaterial(flatIndex) {
  return MATERIAL_GEOMETRY_CLASS.get(flatIndex) ?? 0;
}

export function panelBlockIdForMaterial(flatIndex) {
  return panelBlockIdForGeoClass(geometryClassForMaterial(flatIndex));
}

/**
 * Resolves the specific geometry_id for a material given its neighbor context.
 * For fence-class materials: centered post unless neighbors on both sides.
 * @param {number} flatIndex - material index
 * @param {boolean} hasNeighborBefore - assembly block on the hinge side
 * @param {boolean} hasNeighborAfter - assembly block on the extension side
 */
export function resolveGeometryId(flatIndex, hasNeighborBefore, hasNeighborAfter) {
  const geoClass = MATERIAL_GEOMETRY_CLASS.get(flatIndex) ?? 0;
  if (geoClass === 0) return 0;

  if (geoClass === GEOMETRY_CLASS_FENCE) {
    if (hasNeighborBefore && hasNeighborAfter) return 4;
    if (hasNeighborBefore) return 2;
    if (hasNeighborAfter) return 3;
    return 1;
  }

  if (geoClass === GEOMETRY_CLASS_BARS) {
    if (hasNeighborBefore && hasNeighborAfter) return 5;
    if (hasNeighborBefore) return 10;
    if (hasNeighborAfter) return 11;
    return 9;
  }

  if (geoClass === GEOMETRY_CLASS_PANE) {
    if (hasNeighborBefore && hasNeighborAfter) return 7;
    if (hasNeighborBefore) return 13;
    if (hasNeighborAfter) return 14;
    return 12;
  }

  return geoClass;
}

export function materialToBlockStates(flatIndex, geometryId) {
  return {
    "bigdoors:material_group": Math.floor(flatIndex / 16),
    "bigdoors:material_id": flatIndex % 16,
    "bigdoors:geometry_id": geometryId ?? (MATERIAL_GEOMETRY_CLASS.get(flatIndex) ?? 0),
  };
}

export function blockStatesToMaterial(group, id) {
  return group * 16 + id;
}

/**
 * Compute the full block states for a panel including geometry and rotation.
 * @param {number} matIdx - material index
 * @param {number} geoId - resolved geometry_id
 * @param {number} rotation - panel_rotation (0-3, representing 0/90/180/270 Y degrees)
 */
export function panelBlockStates(matIdx, geoId, rotation, overlay = 0) {
  return {
    "bigdoors:material_group": Math.floor(matIdx / 16),
    "bigdoors:material_id": matIdx % 16,
    "bigdoors:geometry_id": geoId,
    "bigdoors:panel_rotation": rotation,
    "bigdoors:overlay": overlay,
  };
}

export function hasAssemblyBlockAtClosed(assembly, pos) {
  for (const p of assembly.panelPositions) {
    const cp = p.closedPos ?? p;
    if (cp.x === pos.x && cp.y === pos.y && cp.z === pos.z) return true;
  }
  for (const hp of assembly.hingePositions ?? []) {
    if (hp.x === pos.x && hp.y === pos.y && hp.z === pos.z) return true;
  }
  return false;
}

export function resolveVerticalGeometryId(assembly, panelIndex) {
  const panel = assembly.panelPositions[panelIndex];
  const matIdx = panel.materialIndex;
  const pos = panel.closedPos ?? panel;
  const facing = assembly.facing;

  let beforeDir, afterDir;
  if (facing === "north" || facing === "south") {
    beforeDir = "west";
    afterDir = "east";
  } else {
    beforeDir = "north";
    afterDir = "south";
  }

  const bOff = DIR_OFFSETS[beforeDir];
  const aOff = DIR_OFFSETS[afterDir];
  const beforePos = { x: pos.x + bOff.x, y: pos.y + bOff.y, z: pos.z + bOff.z };
  const afterPos = { x: pos.x + aOff.x, y: pos.y + aOff.y, z: pos.z + aOff.z };

  const hasNeighborBefore = hasAssemblyBlockAtClosed(assembly, beforePos);
  const hasNeighborAfter = hasAssemblyBlockAtClosed(assembly, afterPos);
  return resolveGeometryId(matIdx, hasNeighborBefore, hasNeighborAfter);
}

File: java-bigdoors/src/main/java/com/bigdoors/domain/MaterialRegistry.java
package com.bigdoors.domain;

import com.bigdoors.util.Constants;

import java.util.HashMap;
import java.util.Map;

/**
 * Maps between material type IDs and their integer indices.
 * Provides geometry class resolution for exotic panel block types.
 * Pure domain class — no Minecraft imports allowed.
 */
public final class MaterialRegistry {

    private MaterialRegistry() {}

    private static final Map<String, Integer> TYPE_TO_INDEX;

    static {
        TYPE_TO_INDEX = new HashMap<>();
        for (int i = 0; i < Constants.MATERIAL_INDEX.length; i++) {
            if (Constants.MATERIAL_INDEX[i] != null) {
                TYPE_TO_INDEX.put(Constants.MATERIAL_INDEX[i], i);
            }
        }
    }

    public static int indexForTypeId(String typeId) {
        if (typeId == null) return -1;
        Integer idx = TYPE_TO_INDEX.get(typeId);
        return idx != null ? idx : -1;
    }

    public static String typeIdForIndex(int index) {
        if (index < 0 || index >= Constants.MATERIAL_INDEX.length) {
            return null;
        }
        return Constants.MATERIAL_INDEX[index];
    }

    public static boolean isSupportedMaterial(String typeId) {
        if (typeId == null || typeId.isEmpty()) return false;
        return indexForTypeId(typeId) >= 0;
    }

    public static int materialGroupForIndex(int flatIndex) {
        return flatIndex / 16;
    }

    public static int materialIdForIndex(int flatIndex) {
        return flatIndex % 16;
    }

    public static int flatIndexFromGroupAndId(int group, int id) {
        return group * 16 + id;
    }

    public static int geometryClassForMaterial(int flatIndex) {
        return Constants.MATERIAL_GEOMETRY_CLASS.getOrDefault(flatIndex, 0);
    }

    public static String panelBlockIdForMaterial(int flatIndex) {
        return Constants.panelBlockIdForGeoClass(geometryClassForMaterial(flatIndex));
    }

    /**
     * Resolves the specific geometry_id for a material given its neighbor context.
     * Fence-class: solo/before/after/both variants.
     * Bars-class: full/solo/before/after variants.
     * Pane-class: full/solo/before/after variants.
     * Slab-class: returns the geometry class directly (no neighbor variants).
     * Full block: returns 0.
     */
    public static int resolveGeometryId(int flatIndex, boolean hasNeighborBefore, boolean hasNeighborAfter) {
        int geoClass = geometryClassForMaterial(flatIndex);
        if (geoClass == 0) return 0;

        if (geoClass == Constants.GEOMETRY_CLASS_FENCE) {
            if (hasNeighborBefore && hasNeighborAfter) return 4;
            if (hasNeighborBefore) return 2;
            if (hasNeighborAfter) return 3;
            return 1;
        }

        if (geoClass == Constants.GEOMETRY_CLASS_BARS) {
            if (hasNeighborBefore && hasNeighborAfter) return 5;
            if (hasNeighborBefore) return 10;
            if (hasNeighborAfter) return 11;
            return 9;
        }

        if (geoClass == Constants.GEOMETRY_CLASS_PANE) {
            if (hasNeighborBefore && hasNeighborAfter) return 7;
            if (hasNeighborBefore) return 13;
            if (hasNeighborAfter) return 14;
            return 12;
        }

        // Slab and any other class: return the class directly
        return geoClass;
    }

    /**
     * Compute the full block states map for a panel.
     */
    public static Map<String, Integer> panelBlockStates(int matIdx, int geoId, int rotation, int overlay) {
        Map<String, Integer> states = new HashMap<>();
        states.put("bigdoors:material_group", materialGroupForIndex(matIdx));
        states.put("bigdoors:material_id", materialIdForIndex(matIdx));
        states.put("bigdoors:geometry_id", geoId);
        states.put("bigdoors:panel_rotation", rotation);
        states.put("bigdoors:overlay", overlay);
        return states;
    }
}
